# **Importación de conjuntos de datos**

## Objetivos

Al finalizar esta practica, serás capaz de:

- Adquirir datos de diversas maneras.
- Obtener información valiosa utilizando la biblioteca **`Pandas`**.

## Tabla de contenidos

¡Prepárate para explorar los datos como un profesional! Aquí tienes los pasos que seguiremos:

1. Adquisición de Datos
2. Información básica del conjunto de datos

---

¡Ahora es tu turno de comenzar la aventura con **Pandas**! 🚀


# **Adquisición de datos**

Un conjunto de datos es, por lo general, un archivo que contiene información almacenada en uno de varios formatos. Entre los formatos más comunes están: **.csv**, **.json**, **.xlsx**, entre otros. Este conjunto de datos puede estar ubicado en diferentes lugares, como tu computadora local, un servidor, un sitio web o incluso en almacenamiento en la nube. 🌐

Para analizar datos en un notebook de Python, necesitas primero traer el conjunto de datos al entorno del notebook. En esta sección, aprenderás cómo cargar un conjunto de datos en tu **Jupyter Notebook**. 💻

Como ejemplo, utilizaremos el conjunto de datos **Automobile Data**, que está disponible en línea en formato CSV (valores separados por comas). Este será nuestro recurso para practicar cómo leer datos.

### Detalles del conjunto de datos:

- **Fuente de datos**: [Conjunto de datos de automóviles](https://drive.google.com/file/d/1wSJnxg2b86VQGZ-cgt-PJyj4TsZ5djJ9/view?usp=sharing)  
- **Tipo de datos**: CSV  

La biblioteca **Pandas** es una herramienta muy popular y extremadamente útil que te permitirá leer diversos conjuntos de datos y convertirlos en un marco de datos (data frame). ¡No te preocupes! Las plataformas de Jupyter Notebook ya tienen **Pandas** integrado, así que solo necesitas importarlo para comenzar. 🚀

¡Manos a la obra! 🛠️

# Nota sobre instalación de bibliotecas

Si estás utilizando este notebook en un entorno donde no están instaladas las bibliotecas necesarias o si necesitas versiones específicas, puedes instalar las bibliotecas manualmente. Solo necesitas descomentar las líneas a continuación y ejecutarlas:

```python

Descomenta estas líneas si necesitas instalar versiones específicas de las bibliotecas:


In [1]:
# ! mamba install pandas==1.3.3 -y
# ! mamba install numpy=1.21.2 -y

In [2]:
# Importa la biblioteca pandas para manipulación y análisis de datos
import pandas as pd

# Importa la biblioteca numpy para operaciones matemáticas y manejo de arreglos
import numpy as np

## Lectura de Datos

Utilizamos la función `pandas.read_csv()` para leer archivos CSV. En esta practica, que se ejecuta en **Google Colab**, puedes descargar directamente el conjunto de datos desde una URL o trabajar con Google Drive.

El código proporcionado a continuación te permite descargar el conjunto de datos y guardarlo localmente en el entorno de Colab. 🚀


In [3]:
# Importamos las bibliotecas necesarias
import requests

# Define la función para descargar el archivo desde Google Drive
def download_data_from_drive(url, filename):
    """
    Descarga un archivo desde Google Drive y lo guarda localmente en el entorno de Colab.

    Parámetros:
    - url (str): URL de Google Drive del archivo a descargar.
    - filename (str): Nombre con el que se guardará el archivo localmente.
    """
    # Asegúrate de reemplazar 'drive.google.com' por 'uc?export=download' para obtener el archivo correctamente
    file_id = url.split('/d/')[1].split('/')[0]
    download_url = f'https://drive.google.com/uc?export=download&id={file_id}'

    # Realiza la solicitud GET para descargar el archivo
    response = requests.get(download_url)

    # Verifica que la descarga sea exitosa (código de estado 200)
    if response.status_code == 200:
        # Guarda el contenido del archivo en el sistema local
        with open(filename, 'wb') as file:
            file.write(response.content)
        print(f"Archivo descargado y guardado como {filename}")
    else:
        print("Error al descargar el archivo.")

# URL del archivo en Google Drive
file_url = "https://drive.google.com/file/d/1wSJnxg2b86VQGZ-cgt-PJyj4TsZ5djJ9/view?usp=drive_link"
# Nombre del archivo a guardar localmente
filename = "automobile_data.csv"

# Llamamos a la función para descargar el archivo
download_data_from_drive(file_url, filename)

Archivo descargado y guardado como automobile_data.csv


Utiliza el método de **Pandas** `read_csv()` para cargar los datos en un **dataframe**.

In [4]:
# Importamos la biblioteca pandas para manipular y analizar los datos
import pandas as pd

# Cargamos el archivo CSV en un dataframe usando el método 'read_csv' de pandas
df = pd.read_csv('automobile_data.csv')

# Mostramos las primeras filas del dataframe para verificar que los datos se han cargado correctamente
df.head()

3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...  \
0  3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
1  1    ?  alfa-romero  gas  std   two    hatchback  rwd  front  94.5  ...   
2  2  164         audi  gas  std  four        sedan  fwd  front  99.8  ...   
3  2  164         audi  gas  std  four        sedan  4wd  front  99.4  ...   
4  2    ?         audi  gas  std   two        sedan  fwd  front  99.8  ...   

   130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
0  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
1  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
2  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
3  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  
4  136  mpfi  3.19  3.40   8.5  110  5500  19  25  15250  

[5 rows x 26 columns]

Después de leer el conjunto de datos, podemos usar el método `data_frame.head(n)` para ver las primeras n filas del **dataframe**, donde n es un número entero. A diferencia de `data_frame.head(n)`, el método `data_frame.tail(n)` te mostrará las últimas n filas del **dataframe**.

In [5]:
# Mostramos las primeras 5 filas del dataframe
df.head(5)

# Mostramos las últimas 5 filas del dataframe
df.tail(5)

3   ? alfa-romero     gas    std   two convertible  rwd  front   88.6  \
199 -1  95       volvo     gas    std  four       sedan  rwd  front  109.1   
200 -1  95       volvo     gas  turbo  four       sedan  rwd  front  109.1   
201 -1  95       volvo     gas    std  four       sedan  rwd  front  109.1   
202 -1  95       volvo  diesel  turbo  four       sedan  rwd  front  109.1   
203 -1  95       volvo     gas  turbo  four       sedan  rwd  front  109.1   

     ...  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
199  ...  141  mpfi  3.78  3.15   9.5  114  5400  23  28  16845  
200  ...  141  mpfi  3.78  3.15   8.7  160  5300  19  25  19045  
201  ...  173  mpfi  3.58  2.87   8.8  134  5500  18  23  21485  
202  ...  145   idi  3.01  3.40  23.0  106  4800  26  27  22470  
203  ...  141  mpfi  3.78  3.15   9.5  114  5400  19  25  22625  

[5 rows x 26 columns]

## Agregar Encabezados

Echa un vistazo al conjunto de datos. **Pandas** automáticamente asigna encabezados con enteros que comienzan desde 0.

Para describir mejor los datos, puedes agregar un encabezado personalizado. Esta información está disponible en: [Automobile Dataset](https://archive.ics.uci.edu/ml/datasets/Automobile).

Por lo tanto, debes agregar los encabezados manualmente.

Primero, crea una lista llamada "headers" que contenga todos los nombres de las columnas en el orden correspondiente. Luego, usa `dataframe.columns = headers` para reemplazar los encabezados predeterminados con la lista que creaste.

In [6]:
# Lista con los nombres de las columnas (encabezados)
headers = ["symboling", "normalized-losses", "make", "fuel-type", "aspiration",
           "num-of-doors", "body-style", "drive-wheels", "engine-location",
           "wheel-base", "length", "width", "height", "curb-weight",
           "engine-type", "num-of-cylinders", "engine-size", "fuel-system",
           "bore", "stroke", "compression-ratio", "horsepower", "peak-rpm",
           "city-mpg", "highway-mpg", "price"]

# Reemplazamos los encabezados del dataframe con la lista creada
df.columns = headers

# Mostramos las primeras filas del dataframe para verificar los cambios
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          1                 ?  alfa-romero       gas        std          two   
2          2               164         audi       gas        std         four   
3          2               164         audi       gas        std         four   
4          2                 ?         audi       gas        std          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1    hatchback          rwd           front        94.5  ...          152   
2        sedan          fwd           front        99.8  ...          109   
3        sedan          4wd           front        99.4  ...          136   
4        sedan          fwd           front        99.8  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  2.68    3.47               9.0        154      5000       19   
2         mpfi  3.19    3.40              10.0        102      5500       24   
3         mpfi  3.19    3.40               8.0        115      5500       18   
4         mpfi  3.19    3.40               8.5        110      5500       19   

  highway-mpg  price  
0          27  16500  
1          26  16500  
2          30  13950  
3          22  17450  
4          25  15250  

[5 rows x 26 columns]

También puedes ver las primeras 10 entradas del **dataframe** actualizado y notar que los encabezados se han actualizado correctamente.

In [7]:
# Mostramos las primeras 10 filas del dataframe actualizado
df.head(10)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          1                 ?  alfa-romero       gas        std          two   
2          2               164         audi       gas        std         four   
3          2               164         audi       gas        std         four   
4          2                 ?         audi       gas        std          two   
5          1               158         audi       gas        std         four   
6          1                 ?         audi       gas        std         four   
7          1               158         audi       gas      turbo         four   
8          0                 ?         audi       gas      turbo          two   
9          2               192          bmw       gas        std          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1    hatchback          rwd           front        94.5  ...          152   
2        sedan          fwd           front        99.8  ...          109   
3        sedan          4wd           front        99.4  ...          136   
4        sedan          fwd           front        99.8  ...          136   
5        sedan          fwd           front       105.8  ...          136   
6        wagon          fwd           front       105.8  ...          136   
7        sedan          fwd           front       105.8  ...          131   
8    hatchback          4wd           front        99.5  ...          131   
9        sedan          rwd           front       101.2  ...          108   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  2.68    3.47               9.0        154      5000       19   
2         mpfi  3.19    3.40              10.0        102      5500       24   
3         mpfi  3.19    3.40               8.0        115      5500       18   
4         mpfi  3.19    3.40               8.5        110      5500       19   
5         mpfi  3.19    3.40               8.5        110      5500       19   
6         mpfi  3.19    3.40               8.5        110      5500       19   
7         mpfi  3.13    3.40               8.3        140      5500       17   
8         mpfi  3.13    3.40               7.0        160      5500       16   
9         mpfi  3.50    2.80               8.8        101      5800       23   

  highway-mpg  price  
0          27  16500  
1          26  16500  
2          30  13950  
3          22  17450  
4          25  15250  
5          25  17710  
6          25  18920  
7          20  23875  
8          22      ?  
9          29  16430  

[10 rows x 26 columns]

Ahora, necesitamos reemplazar el símbolo "?" por NaN para que el método `dropna()` pueda eliminar los valores faltantes.

In [8]:
# Reemplazamos los símbolos "?" por NaN
df.replace("?", float("nan"), inplace=True)

# Verificamos si el reemplazo fue exitoso mostrando las primeras filas
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          1               NaN  alfa-romero       gas        std          two   
2          2               164         audi       gas        std         four   
3          2               164         audi       gas        std         four   
4          2               NaN         audi       gas        std          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1    hatchback          rwd           front        94.5  ...          152   
2        sedan          fwd           front        99.8  ...          109   
3        sedan          4wd           front        99.4  ...          136   
4        sedan          fwd           front        99.8  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  2.68    3.47               9.0        154      5000       19   
2         mpfi  3.19    3.40              10.0        102      5500       24   
3         mpfi  3.19    3.40               8.0        115      5500       18   
4         mpfi  3.19    3.40               8.5        110      5500       19   

  highway-mpg  price  
0          27  16500  
1          26  16500  
2          30  13950  
3          22  17450  
4          25  15250  

[5 rows x 26 columns]

Puedes eliminar los valores faltantes de la columna "price" de la siguiente manera:

In [9]:
# Eliminamos las filas con valores faltantes en la columna "price"
df.dropna(subset=["price"], inplace=True)

# Verificamos si las filas con valores faltantes fueron eliminadas
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          1               NaN  alfa-romero       gas        std          two   
2          2               164         audi       gas        std         four   
3          2               164         audi       gas        std         four   
4          2               NaN         audi       gas        std          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1    hatchback          rwd           front        94.5  ...          152   
2        sedan          fwd           front        99.8  ...          109   
3        sedan          4wd           front        99.4  ...          136   
4        sedan          fwd           front        99.8  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  2.68    3.47               9.0        154      5000       19   
2         mpfi  3.19    3.40              10.0        102      5500       24   
3         mpfi  3.19    3.40               8.0        115      5500       18   
4         mpfi  3.19    3.40               8.5        110      5500       19   

  highway-mpg  price  
0          27  16500  
1          26  16500  
2          30  13950  
3          22  17450  
4          25  15250  

[5 rows x 26 columns]

In [10]:
# Eliminamos las filas donde la columna "price" tiene valores NaN
df.dropna(subset=["price"], axis=0, inplace=True)

# Verificamos si el dataframe se ha actualizado correctamente
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          1               NaN  alfa-romero       gas        std          two   
2          2               164         audi       gas        std         four   
3          2               164         audi       gas        std         four   
4          2               NaN         audi       gas        std          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1    hatchback          rwd           front        94.5  ...          152   
2        sedan          fwd           front        99.8  ...          109   
3        sedan          4wd           front        99.4  ...          136   
4        sedan          fwd           front        99.8  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  2.68    3.47               9.0        154      5000       19   
2         mpfi  3.19    3.40              10.0        102      5500       24   
3         mpfi  3.19    3.40               8.0        115      5500       18   
4         mpfi  3.19    3.40               8.5        110      5500       19   

  highway-mpg  price  
0          27  16500  
1          26  16500  
2          30  13950  
3          22  17450  
4          25  15250  

[5 rows x 26 columns]

Aquí, `axis=0` significa que se eliminarán las filas completas donde se encuentre un valor **NaN** en la columna 'price'.

Ahora, has leído exitosamente el conjunto de datos sin procesar y agregado los encabezados correctos al **dataframe**.

Encontrar los nombres de las columnas del **dataframe**.


In [11]:
# Obtenemos y mostramos los nombres de las columnas del dataframe
df.columns

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')

## Guardar conjunto de datos

De manera similar, **Pandas** te permite guardar el conjunto de datos en formato CSV. Usando el método `dataframe.to_csv()`, puedes especificar la ruta y el nombre del archivo dentro de los paréntesis.

Por ejemplo, si guardas el **dataframe** `df` como `automobile.csv` en tu máquina local, puedes utilizar la siguiente sintaxis, donde `index=False` significa que los nombres de las filas no se escribirán.

In [12]:
# Guardamos el dataframe como un archivo CSV sin los nombres de las filas
df.to_csv('automobile.csv', index=False)

## Leer/Guardar Otros Formatos de Datos

También puedes leer y guardar otros formatos de archivo. Puedes usar funciones similares a `pd.read_csv()` y `df.to_csv()` para otros formatos de datos. Las funciones se enumeran en la siguiente tabla:

| Formato de Datos | Leer | Guardar |
|------------------|------|---------|
| csv              | `pd.read_csv()` | `df.to_csv()` |
| json             | `pd.read_json()` | `df.to_json()` |
| excel            | `pd.read_excel()` | `df.to_excel()` |
| hdf              | `pd.read_hdf()` | `df.to_hdf()` |
| sql              | `pd.read_sql()` | `df.to_sql()` |
| ...              | ...  | ... |


## Información básica del conjunto de datos

Después de cargar los datos en el **dataframe** de Pandas, es momento de explorar el conjunto de datos.

Existen varias formas de obtener información esencial sobre los datos que te ayudarán a comprenderlos mejor.

## Tipos de Datos

Los datos tienen una variedad de tipos.

Los principales tipos almacenados en los **dataframes** de Pandas son: **object**, **float**, **int**, **bool** y **datetime64**. Para aprender mejor sobre cada atributo, siempre debes conocer el tipo de dato de cada columna. En Pandas:

Devuelve una serie con el tipo de dato de cada columna.
`df.dtypes`

In [14]:
# Obtenemos y mostramos los tipos de datos de cada columna
df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

Como se muestra arriba, puedes ver claramente que el tipo de dato de "symboling" y "curb-weight" es `int64`, "normalized-losses" es `object`, y "wheel-base" es `float64`, entre otros.

Estos tipos de datos pueden cambiarse; aprenderás cómo hacerlo en un módulo posterior.

## Describe

Si deseas obtener un resumen estadístico de cada columna, como el conteo, el valor medio de la columna, la desviación estándar de la columna, etc., utiliza el método `describe`:

In [15]:
# Obtenemos un resumen estadístico de cada columna
df.describe()

symboling  wheel-base      length       width      height  \
count  200.000000  200.000000  200.000000  200.000000  200.000000   
mean     0.830000   98.848000  174.228000   65.898000   53.791500   
std      1.248557    6.038261   12.347132    2.102904    2.428449   
min     -2.000000   86.600000  141.100000   60.300000   47.800000   
25%      0.000000   94.500000  166.675000   64.175000   52.000000   
50%      1.000000   97.000000  173.200000   65.500000   54.100000   
75%      2.000000  102.400000  183.500000   66.675000   55.525000   
max      3.000000  120.900000  208.100000   72.000000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   200.000000   200.000000         200.000000  200.000000   200.000000  
mean   2555.705000   126.860000          10.170100   25.200000    30.705000  
std     518.594552    41.650501           4.014163    6.432487     6.827227  
min    1488.000000    61.000000           7.000000   13.000000    16.000000  
25%    2163.000000    97.750000           8.575000   19.000000    25.000000  
50%    2414.000000   119.500000           9.000000   24.000000    30.000000  
75%    2928.250000   142.000000           9.400000   30.000000    34.000000  
max    4066.000000   326.000000          23.000000   49.000000    54.000000

Este método proporcionará varias estadísticas resumen, excluyendo los valores `NaN` (No es un Número).

Esto muestra el resumen estadístico de todas las columnas de tipo numérico (**int**, **float**).

Por ejemplo, el atributo "symboling" tiene 205 registros, el valor medio de esta columna es 0.83, la desviación estándar es 1.25, el valor mínimo es -2, el percentil 25 es 0, el percentil 50 es 1, el percentil 75 es 2 y el valor máximo es 3.

Sin embargo, ¿qué pasa si también deseas revisar todas las columnas, incluidas aquellas de tipo **object**?

Puedes agregar un argumento `include = "all"` dentro del paréntesis. Inténtalo de nuevo.

In [16]:
# Obtenemos un resumen estadístico de todas las columnas, incluidas las de tipo "object"
df.describe(include="all")

symboling normalized-losses    make fuel-type aspiration  \
count   200.000000               164     200       200        200   
unique         NaN                51      22         2          2   
top            NaN               161  toyota       gas        std   
freq           NaN                11      32       180        164   
mean      0.830000               NaN     NaN       NaN        NaN   
std       1.248557               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           198        200          200             200  200.000000  ...   
unique            2          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            113         94          118             197         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.848000  ...   
std             NaN        NaN          NaN             NaN    6.038261  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

        engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
count    200.000000          200   196     196        200.000000        198   
unique          NaN            8    38      36               NaN         58   
top             NaN         mpfi  3.62    3.40               NaN         68   
freq            NaN           91    23      19               NaN         19   
mean     126.860000          NaN   NaN     NaN         10.170100        NaN   
std       41.650501          NaN   NaN     NaN          4.014163        NaN   
min       61.000000          NaN   NaN     NaN          7.000000        NaN   
25%       97.750000          NaN   NaN     NaN          8.575000        NaN   
50%      119.500000          NaN   NaN     NaN          9.000000        NaN   
75%      142.000000          NaN   NaN     NaN          9.400000        NaN   
max      326.000000          NaN   NaN     NaN         23.000000        NaN   

        peak-rpm    city-mpg highway-mpg  price  
count        198  200.000000  200.000000    200  
unique        22         NaN         NaN    185  
top         5500         NaN         NaN  16500  
freq          36         NaN         NaN      2  
mean         NaN   25.200000   30.705000    NaN  
std          NaN    6.432487    6.827227    NaN  
min          NaN   13.000000   16.000000    NaN  
25%          NaN   19.000000   25.000000    NaN  
50%          NaN   24.000000   30.000000    NaN  
75%          NaN   30.000000   34.000000    NaN  
max          NaN   49.000000   54.000000    NaN  

[11 rows x 26 columns]

Ahora proporciona el resumen estadístico de todas las columnas, incluidas las de tipo **object**.

Ahora puedes ver cuántos valores únicos hay, cuál es el valor más frecuente y la frecuencia de ese valor en las columnas de tipo **object**.

Algunos valores en la tabla anterior muestran "NaN". Esos números no están disponibles para un tipo de columna en particular.

Puedes seleccionar las columnas de un **dataframe** indicando el nombre de cada columna. Por ejemplo, puedes seleccionar las tres columnas de la siguiente manera:

```python
dataframe[['column 1', 'column 2', 'column 3']]
```

In [19]:
# Seleccionar las columnas 'length' y 'compression-ratio' y obtener un resumen estadístico
df[['length', 'compression-ratio']].describe()

length  compression-ratio
count  200.000000         200.000000
mean   174.228000          10.170100
std     12.347132           4.014163
min    141.100000           7.000000
25%    166.675000           8.575000
50%    173.200000           9.000000
75%    183.500000           9.400000
max    208.100000          23.000000

### Información

También puedes usar otro método para revisar tu conjunto de datos:
### `info`

Proporciona un resumen conciso de tu **dataframe**.

Este método imprime información sobre el **dataframe**, incluyendo el tipo de índice, las columnas, los valores no nulos y el uso de

In [20]:
# Obtener información general sobre el dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 200 entries, 0 to 203
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          200 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               200 non-null    object 
 3   fuel-type          200 non-null    object 
 4   aspiration         200 non-null    object 
 5   num-of-doors       198 non-null    object 
 6   body-style         200 non-null    object 
 7   drive-wheels       200 non-null    object 
 8   engine-location    200 non-null    object 
 9   wheel-base         200 non-null    float64
 10  length             200 non-null    float64
 11  width              200 non-null    float64
 12  height             200 non-null    float64
 13  curb-weight        200 non-null    int64  
 14  engine-type        200 non-null    object 
 15  num-of-cylinders   200 non-null    object 
 16  engine-size        200 non-null

¡Excelente! Ha completado el cuaderno de Introducción. 🎉